### PR TITLE
(GH-399) Multi-monitor support in Set-TaskbarOptions

### PR DESCRIPTION
**Pull request by HolisticDeveloper**
Adds support for setting the options for multiple displays in Windows 10.

### CODE REVIEW

File: Boxstarter.WinConfig/Set-TaskbarOptions.ps1 [hunk 1]
```diff
@@ -55,8 +67,16 @@ Turn off always show all icons in the notification area
...
+				$Combine,
+        [Parameter(ParameterSetName='MultiMonitorOn')]
+        [switch]$MultiMonitorOn,
+        [Parameter(ParameterSetName='MultiMonitorOff')]
+        [switch]$MultiMonitorOff,
+        [ValidateSet('All', 'MainAndOpen', 'Open')]
+        $MultiMonitorMode,
+        [ValidateSet('Always','Full','Never')]
+        $MultiMonitorCombine
+    )
```
Comment (pauby) [inline Boxstarter.WinConfig/Set-TaskbarOptions.ps1]: Should the switches `MultiMonitorMode` and `MultiMonitorCombine` not be part of the `MultiMonitorOn` ParameterSet as they are only valid when `MultiMonitorOn` is used?

File: Boxstarter.WinConfig/Set-TaskbarOptions.ps1 [hunk 2]
```diff
@@ -85,7 +105,29 @@ Turn off always show all icons in the notification area
 			"Full" { Set-ItemProperty $key TaskbarGlomLevel 1 }
 			"Never" { Set-ItemProperty $key TaskbarGlomLevel 2 }
 		}
```
Comment (pauby) [inline Boxstarter.WinConfig/Set-TaskbarOptions.ps1]: Can we use the full parameter names for cmdlets / functions so there is no confusion? I appreciate this hasn't been done in the function but going forward this is a standard we are working towards.

**Comment by pauby** [comment]
@HolisticDeveloper Thanks for this PR. Hopefully we can get it merged soon.

File: Boxstarter.WinConfig/Set-TaskbarOptions.ps1 (hunk 1 above)
Comment (HolisticDeveloper) [reply Boxstarter.WinConfig/Set-TaskbarOptions.ps1]: How do you envision that working on the command line? With the current PR you can do this:

```
Set-TaskbarOptions -MultiMonitorOn -MultiMonitorMode All -MultiMonitorCombine Always
```

File: Boxstarter.WinConfig/Set-TaskbarOptions.ps1 (hunk 2 above)
Comment (HolisticDeveloper) [reply Boxstarter.WinConfig/Set-TaskbarOptions.ps1]: To be clear (as I am fairly new with PowerShell), you're suggesting changing the call like this?

```
Set-ItemProperty -Path $key -Name TaskbarGlomLevel -Value 1
```

File: Boxstarter.WinConfig/Set-TaskbarOptions.ps1 (hunk 1 above)
Comment (pauby) [reply Boxstarter.WinConfig/Set-TaskbarOptions.ps1]: You can also do `Set-TaskbarOptions -MultiMonitorOff -MultiMonitorMode All -MultiMonitorCombine Always`? That doesn't make sense or does it and I'm missing something?

File: Boxstarter.WinConfig/Set-TaskbarOptions.ps1 (hunk 2 above)
Comment (pauby) [reply Boxstarter.WinConfig/Set-TaskbarOptions.ps1]: Yeah, that's what I was looking for. If we use the full cmdlet name in scripts there is no ambiguity of what's meant.

File: Boxstarter.WinConfig/Set-TaskbarOptions.ps1 (hunk 1 above)
Comment (HolisticDeveloper) [reply Boxstarter.WinConfig/Set-TaskbarOptions.ps1]: Ok, I read more about `ParameterSet` and now I understand what you're asking.

**Comment by HolisticDeveloper** [comment]
@pauby Let me know if I need to take further action. At this point I'm just waiting for you to have bandwidth to review the changes I made.

**Comment by pauby** [comment]
@holisticdeveloper I'm going to try and take a look at the PR's this weekend.

**Comment by pauby** [comment]
@HolisticDeveloper Other than the blank lines between the parameter declarations I think we're good to go.

**Comment by HolisticDeveloper** [comment]
@pauby No problem, I can update that. I was trying to follow the existing style in the script. Are you asking to change (for example):
```
        [Parameter(ParameterSetName='MultiMonitorOn')]
        [switch]$MultiMonitorOn,
        [Parameter(ParameterSetName='MultiMonitorOn')]
        [ValidateSet('All', 'MainAndOpen', 'Open')]
        $MultiMonitorMode,
        [Parameter(ParameterSetName='MultiMonitorOn')]
        [ValidateSet('Always','Full','Never')]
        $MultiMonitorCombine
```
to
```
        [Parameter(ParameterSetName='MultiMonitorOn')]
        [switch]$MultiMonitorOn,

        [Parameter(ParameterSetName='MultiMonitorOn')]
        [ValidateSet('All', 'MainAndOpen', 'Open')]
        $MultiMonitorMode,

        [Parameter(ParameterSetName='MultiMonitorOn')]
        [ValidateSet('Always','Full','Never')]
        $MultiMonitorCombine
```
Is there a style guide I can reference that would have these and other guidelines, so I can make sure to avoid running into this for future contributions? I didn't see anything in CONTRIBUTING.md that would address it...

**Comment by HolisticDeveloper** [comment]
@pauby Can you confirm the above example is what you're asking for?

**Comment by pauby** [comment]
@HolisticDeveloper Apologies for the VERY late reply. That's exactly what I'm looking for. In general the code would follow the existing code. However, we're trying to establish some style guidelines so I'm going to add these to the CONTRIBUTING.md.

**Comment by pauby** [comment]
@HolisticDeveloper See https://github.com/chocolatey/boxstarter/pull/408 for PR to update some of the guidelines. This is a work in progress so any suggestions welcome.

**Comment by HolisticDeveloper** [comment]
@pauby As I was going through to update the formatting, I had a couple of extra style questions.

1. The switch parameters all have the `[switch]` attribute on the same line as the parameter name. Typically I would expect one attribute per line with the parameter itself on its own line. Which is preferred?

    ```ps
    [Parameter(ParameterSetName='lockhide')]
    [Parameter(ParameterSetName='locknohide')]
    [switch]$Lock,
    ```
    or
    ```ps
    [Parameter(ParameterSetName='lockhide')]
    [Parameter(ParameterSetName='locknohide')]
    [switch]
    $Lock,
    ```
2. There are some inconsistencies in spacing and brace placement with `if` and `switch` statements. It seems the most common is to have the opening brace at the end of the line with the `if` statement. Also, I would typically put spaces around the condition. Please confirm.

    Existing code:
    ```ps
    if($Lock)
    {
        Set-ItemProperty -Path $key -Name TaskbarSizeMove -Value 0
    }
    ````
    Preferred? (note spaces around `($Lock)`
    ```ps
    if ($Lock) {
        Set-ItemProperty -Path $key -Name TaskbarSizeMove -Value 0
    }
    ```

File: Boxstarter.WinConfig/Set-TaskbarOptions.ps1 [hunk 3]
```diff
@@ -57,17 +57,17 @@ Changes the Taskbar Icon combination style for non-primary displays.  Valid inpu
...
-				[switch]$AlwaysShowIconsOn,
-				[Parameter(ParameterSetName='AlwaysShowIconsOff')]
-				[switch]$AlwaysShowIconsOff,
-				[ValidateSet('Small','Large')]
-				$Size,
-				[ValidateSet('Top','Left','Bottom','Right')]
-				$Dock,
-				[ValidateSet('Always','Full','Never')]
-				$Combine,
+        [switch]$NoAutoHide,
```
[the remark this answers is not in the document]
Comment (HolisticDeveloper) [reply Boxstarter.WinConfig/Set-TaskbarOptions.ps1]: Done, see update.

**Comment by pauby** [comment]
Thanks for your questions. To answer:

1. Switch parameters - I prefer option 2. It takes up more space but is more readable. It should be something else to add to the coding guidelines.

2. Bracing style - this is actually something we haven't decided on yet. However, option 2 is preferred as spaces are good 😄

**Comment by HolisticDeveloper** [comment]
Ok, should be good to go with my latest changes then.

**Comment by pauby** [comment]
@mwallner Can you review this PR when you are free?

**Comment by HolisticDeveloper** [comment]
@pauby what is the next step?

**Comment by pauby** [comment]
Awesome work! Thanks for your contribution @HolisticDeveloper !

**Pull request merged**